third_party/move/move-compiler-v2/src/inliner.rs

```diff
@@ -0,0 +1,1184 @@
+//      { let params=actuals; body } used for both lambda inlining and inline function inlining.
+//
+//   - rewrite_pattern_vector is a helper for `rewrite_pattern`
+//
+
+// TODO:
+// - add a InlinedCall() node that represents an inlined call, so that an inlined Return() has a place to go. [later]
+//   - OR: if an inlined fn has a return, then error out.
+// - add a simplifier that simplifies certain code constructs, e.g.
+//   - InlinedCall() with no nested Return() can be flattened.
+//   - others?
+// - do we need to insert FreezeRef in actual args if formal param is &T but arg is &mod T ?
+// - do we need to do anything about abilities?
+// - if lambda parameters also may be referred to by temporaries, then `rewrite_invoke` might need to call yet
+//   another `ExpRewriterFunctions` implementation to do that.
+
+use codespan_reporting::diagnostic::Severity;
+use itertools::chain;
+use move_model::{
+    ast::{Exp, ExpData, Operation, Pattern, TempIndex},
+    exp_rewriter::ExpRewriterFunctions,
+    model::{FunId, GlobalEnv, Loc, NodeId, Parameter, QualifiedId, TypeParameter},
+    symbol::Symbol,
+    ty::{ReferenceKind, Type},
+};
+use std::{
+    collections::{BTreeMap, BTreeSet},
+    fmt::Debug,
+    iter::{zip, IntoIterator, Iterator},
+    ops::Deref,
+    vec::Vec,
+};
+
+type QualifiedFunId = QualifiedId<FunId>;
+type CallSiteLocations = BTreeMap<(QualifiedFunId, QualifiedFunId), BTreeSet<NodeId>>;
+
+// ======================================================================================
+// Entry
+
+// Run inlining on current program's AST.  For each function which is target of the compilation,
+// visit that function body and inline any calls to functions marked as "inline".
+// debug flag is temporary until we sort out a way to attach debug to GlobalEnv.
+pub fn run_inlining(env: &mut GlobalEnv, debug: bool) {
+    // Start with targets.
+    let mut todo = BTreeSet::new();
+    // While we're iterating, error on any target inline functions lacking a body to inline.
+    let mut has_error: bool = false;
+    for module in env.get_modules() {
+        if module.is_target() {
+            for func in module.get_functions() {
+                let id = func.get_qualified_id();
+                if func.is_inline() {
+                    let func_def = func.get_def();
+                    if func_def.deref().is_none() {
+                        has_error = true;
+                        let func_loc = func.get_loc();
+                        let func_name = func.get_name_str();
+                        if func.is_native() {
+                            let msg = format!("Inline function {} must not be native", func_name);
+                            env.diag(Severity::Error, &func_loc, &msg);
+                        } else {
+                            let msg = format!(
+                                "ICE: No body found for non-native inline function {}",
```

Comment:
There should be no panics.  But, fine.

Reply:
I'm not saying to panic. I meant don't use the ICE prefix because it is usually used for panics. Instead use what we are already doing in many other places with Severity::Bug in v2 tree (I don't know what it is -- I think '[bug]' -- but you can easily find).